### ISSUE TITLE
HVDAllReduce stalls with Missing Ranks during Training

### ISSUE THREAD

**Issue by amrragab8080** [armandmcqueen/tensorpack-mask-rcnn#47]
During training (2 node - 16 GPU p3dn.24xl OpenMPI 4.0 + EFA) I am seeing a situation where HVDALLReduce callback has missing ranks at random points in training.
`[0517 00:57:43 @param.py:161] [HyperParamSetter] At global_step=1, learning_rate changes from 0.003300 to 0.003307
 53%|#####2    |993/1875[12:30<09:18, 1.58it/s]WARNING: One or more tensors were submitted to be reduced, gathered or broadcasted by subset of ranks and are waiting for remainder of ranks for more than 60 seconds. This may indicate that different ranks are trying to submit different tensors or that only subset of ranks is submitting tensors, which will cause deadlock. 
Stalled ops:
HVDAllReduce/HorovodAllreduce_truediv_25_0 [missing ranks: 0, 1, 2, 3, 4, 5, 6, 7]
HVDAllReduce/HorovodAllreduce_truediv_26_0 [missing ranks: 0, 1, 2, 3, 4, 5, 6, 7]`

I am debugging the error currently so I am not expecting to fix anything - however I did find that increasing `HOROVOD_CYCLE_TIME` did improve the failure rate, perhaps the tensor fusion is causing some stability issues.

**Comment by amrragab8080** [comment]
After some further testing - related to `HOROVOD_CYCLE_TIME` I was able to complete a training by completely disabling Tensor Fusion setting `HOROVOD_FUSION_THRESHOLD=0`.

**Comment by armandmcqueen** [comment]
That's interesting. Horovod training is a little unstable for this codebase, but the failure rate is very low and it generally manifests as a segfault. I've never seen stability issues caused by `HOROVOD_FUSION_THRESHOLD` or `HOROVOD_CYCLE_TIME` before. 

If you can share the training command you used and/or the mpirun log folder, we can take a look and see if anything stands out from normal.

**Comment by armandmcqueen** [comment]
We have tested EFA and did not experience any stability issues. Feel free to reopen if its reappears

**Comment by amrragab8080** [comment]
Can you please provide more detail such as mpi launch command for example

**Comment by armandmcqueen** [comment]
@YangFei1990

**Comment by YangFei1990** [comment]
@amrragab8080 To run with EFA, we use the command in https://github.com/armandmcqueen/tensorpack-mask-rcnn/blob/master/infra/ami/train_efa.sh
Please let me know if it solved your problem.

**Comment by amrragab8080** [comment]
@YangFei1990 if you are using this:
`--mca plm_rsh_no_tree_spawn 1 -bind-to none -map-by slot -mca pml ob1 \
-mca btl_vader_single_copy_mechanism none \
--mca btl tcp,self \`
This call stack indicates you are using OB1 and Vader which indicates you are still just using TCP backends. So you are not using EFA at all. To force the use of EFA (and reproduce the problem I am seeing) you need to pass this to the call:
`-mca pml cm -mca mtl ofi`
Let me know if that causes the freeze when you run with those directives

**Comment by YangFei1990** [comment]
@amrragab8080 I'm sure that the EFA is enabled since I read “Selected Provider is efa;ofi_rxr” and “Forcing AWS OFI ndev 8” from the NCCL debug info.

**Comment by armandmcqueen** [comment]
We use those flags in our normal training - they may just have been copied and are not actually necessary, I'm not sure. 

We also don't really need MPI to use EFA. The vast majority of network traffic happens via NCCL and as long as NCCL is using EFA, we are getting the bandwidth benefits

**Comment by amrragab8080** [comment]
Correct I see the efa;ofi_rxr when I run the nccl-tests. The inclusion of

```
-x FI_OFI_RXR_RX_COPY_UNEXP=1 \
-x FI_OFI_RXR_RX_COPY_OOO=1 \
-x FI_EFA_MR_CACHE_ENABLE=1 \
-x FI_OFI_RXR_INLINE_MR_ENABLE=1 \
```

Fixes the issue
Thanks
